cmake/FindIlmBase.cmake

```diff
@@ -0,0 +1,337 @@
+  The IexMath library target.
+``IlmBase::IlmThread``
+  The IlmThread library target.
+``IlmBase::Imath``
+  The Imath library target.
+
+Result Variables
+^^^^^^^^^^^^^^^^
+
+This will define the following variables:
+
+``IlmBase_FOUND``
+  True if the system has the IlmBase library.
+``IlmBase_VERSION``
+  The version of the IlmBase library which was found.
+``IlmBase_INCLUDE_DIRS``
+  Include directories needed to use IlmBase.
+``IlmBase_LIBRARIES``
+  Libraries needed to link to IlmBase.
+``IlmBase_LIBRARY_DIRS``
+  IlmBase library directories.
+``IlmBase_{COMPONENT}_FOUND``
+  True if the system has the named IlmBase component.
+
+Cache Variables
+^^^^^^^^^^^^^^^
+
+The following cache variables may also be set:
+
+``IlmBase_INCLUDE_DIR``
+  The directory containing ``IlmBase/config-auto.h``.
+``IlmBase_{COMPONENT}_LIBRARY``
+  Individual component libraries for IlmBase
+``IlmBase_{COMPONENT}_DLL``
+  Individual component dlls for IlmBase on Windows.
+
+Hints
+^^^^^
+
+Instead of explicitly setting the cache variables, the following variables
+may be provided to tell this module where to look.
+
+``ILMBASE_ROOT``
+  Preferred installation prefix.
+``ILMBASE_INCLUDEDIR``
+  Preferred include directory e.g. <prefix>/include
+``ILMBASE_LIBRARYDIR``
+  Preferred library directory e.g. <prefix>/lib
+``SYSTEM_LIBRARY_PATHS``
+  Paths appended to all include and lib searches.
+
+#]=======================================================================]
+
+# Support new if() IN_LIST operator
+IF ( POLICY CMP0057 )
+  CMAKE_POLICY ( SET CMP0057 NEW )
+ENDIF ()
+
+MARK_AS_ADVANCED (
+  IlmBase_INCLUDE_DIR
+  IlmBase_LIBRARY
+)
+
+SET ( _ILMBASE_COMPONENT_LIST
+  Half
+  Iex
+  IexMath
+  IlmThread
+  Imath
+  )
+
+IF ( IlmBase_FIND_COMPONENTS )
+  SET ( ILMBASE_COMPONENTS_PROVIDED TRUE )
+  SET ( _IGNORED_COMPONENTS "" )
+  FOREACH ( COMPONENT ${IlmBase_FIND_COMPONENTS} )
+    IF ( NOT ${COMPONENT} IN_LIST _ILMBASE_COMPONENT_LIST )
+      LIST ( APPEND _IGNORED_COMPONENTS ${COMPONENT} )
+    ENDIF ()
+  ENDFOREACH()
+
+  IF ( _IGNORED_COMPONENTS )
+    MESSAGE ( STATUS "Ignoring unknown components of IlmBase:" )
+    FOREACH ( COMPONENT ${_IGNORED_COMPONENTS} )
+      MESSAGE ( STATUS "  ${COMPONENT}" )
+    ENDFOREACH ()
+    LIST ( REMOVE_ITEM IlmBase_FIND_COMPONENTS ${_IGNORED_COMPONENTS} )
+  ENDIF ()
+ELSE ()
+  SET ( ILMBASE_COMPONENTS_PROVIDED FALSE )
+  SET ( IlmBase_FIND_COMPONENTS ${_ILMBASE_COMPONENT_LIST} )
+ENDIF ()
+
+# Append ILMBASE_ROOT or $ENV{ILMBASE_ROOT} if set (prioritize the direct cmake var)
+SET ( _ILMBASE_ROOT_SEARCH_DIR "" )
+
+IF ( ILMBASE_ROOT )
+  LIST ( APPEND _ILMBASE_ROOT_SEARCH_DIR ${ILMBASE_ROOT} )
+ELSE ()
+  SET ( _ENV_ILMBASE_ROOT $ENV{ILMBASE_ROOT} )
+  IF ( _ENV_ILMBASE_ROOT )
+    LIST ( APPEND _ILMBASE_ROOT_SEARCH_DIR ${_ENV_ILMBASE_ROOT} )
+  ENDIF ()
+ENDIF ()
+
+# Additionally try and use pkconfig to find IlmBase
+
+FIND_PACKAGE ( PkgConfig )
+PKG_CHECK_MODULES ( PC_IlmBase QUIET IlmBase )
```

Comment:
Ran into a bit of a small issue with this logic. Many systems have old versions of IlmBase/OpenEXR pre-installed in /usr/include/OpenEXR and /usr/lib64. On mine I have IlmBase 1.0.3 and OpenEXR 1.7.1.

I used USE_SYSTEM_LIBRARY_PATHS=OFF to ensure it didn't pick up these old paths, however this PKG_CHECK_MODULES line still finds /usr/lib64/pkgconfig/IlmBase.pc which adds those path to the subsequent FIND_PATH() routine. Right now this gives a slightly obscure message due to the fact that IlmBaseConfig.h is so old and the version string used to be different:

> Could NOT find IlmBase: Found unsuitable version ".", but required is at
>   least "2.2" (found /usr/include/OpenEXR)

I can't seem to find a nice platform-independent way of disabling "system" pkg-config search paths? Perhaps that means ignoring any system paths that are returned by pkg-config, rather than blinding adding them to the include search path, I'm not sure.

Reply:
I don't think this requires any changes with the way the pkg-config/system paths work - just need to handle the case where the version isn't parsed correctly (assuming that it always will be for newer versions). I could update the regex to parse the `#define ILMBASE_VERSION_STRING` which is defined as far back as 1.0.3. Or instead just elevate a more appropriate error if the version can't be retrieved.

Reply:
In the case where the installed system library is older than our minimum requirements, that's fine because it's pretty clear what's going on, but I would like to be able to disable the use of system headers and ensure that it doesn't accidentally pick up the wrong library if that weren't to be the case.

Reply:
But the pkg-config paths are also only a fallback, the same as the system paths. If you don't provide direct locations using the include/library or root variables then the alternative would it that it finds nothing? And if you're not providing a direct location then picking up the pkg-config path probably is the "correct" library. 

Or is it more that you want to make absolutely certain that cmake has found the library you expect it to find and hasn't slightly failed? If that's the case then theres probably a few things we could do:

 - If include/library or root paths are given, only use those and don't adhere to any system or pkg-config results. I think find modules provided with cmake do what we're currently doing.
 - Add an option to disable pkg-config searches
 - Improve the verbosity and frequency of status/error messages (i.e. print the search path it ended up using/print a warning if it couldn't find a path using your provided variables and had to fall back). Interestingly it would have printed its chosen path had it actually successfully retrieved the version information.

Reply:
> * Add an option to disable pkg-config searches
You make some good points - I think that this is probably the best solution. At least by being able to disable all the fallbacks you won't get caught out by the wrong library being picked up by accident.

Reply:
Added to the list of things to do, holding off for this PR

Reply:
Sounds good!